### PR TITLE
System Notifications - Data models changes [SCI-2952]

### PR DESCRIPTION
**Pull request by urbanrotnik**
Jira ticket: [SCI-2952](https://biosistemika.atlassian.net/browse/SCI-2952)

**ToDo:**
Add data models for System Notifications

**What's done:** 
Added models and specs. 

**Note**
@Ducz0r 
Refer to JIRA ticket... `Maybe - not entirely sure - we would also like to update the User's Settings JSONB...` 
Should we or not? :)

### CODE REVIEW

File: db/migrate/20190125122852_create_system_notifications.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class CreateSystemNotifications < ActiveRecord::Migration[5.1]
+  def change
+    create_table :system_notifications do |t|
+      t.uuid :UUID
```
[the remark this answers is not in the document]
Comment (urbanrotnik) [reply db/migrate/20190125122852_create_system_notifications.rb]: Sure, I don't know why I typed it uppercase... :red_lips: :)

File: db/migrate/20190125123107_create_user_system_notifications.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class CreateUserSystemNotifications < ActiveRecord::Migration[5.1]
+  def change
+    create_table :user_system_notifications do |t|
+      t.references :user, foreign_key: true
+      t.references :system_notification, foreign_key: true
+      t.datetime :seen, index: true
+      t.datetime :read, index: true
```
[the remark this answers is not in the document]
Comment (urbanrotnik) [reply db/migrate/20190125123107_create_user_system_notifications.rb]: Yes, this is what I thought. And have extra info when been seen/read. 
But... Hm... 
If you don't see any sense to have this info, we should change it to `boolean`.  If it stays as timestamp, probably suffix `_at` should be added?

File: app/models/system_notification.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class SystemNotification < ApplicationRecord
+  has_many :user_system_notifications
+  has_many :users, through: :user_system_notifications
+
+  validates :title, :modal_title, :modal_body, :original_created_at,
```
[the remark this answers is not in the document]
Comment (urbanrotnik) [reply app/models/system_notification.rb]: Agree.

**Comment by Ducz0r** [comment]
Regarding refactoring `User.settings`, maybe it's not worth it. We can keep the current structure of the `jsonb` field, so we keep this "structure":

```
{
    ...,
    notifications_settings: {
        ...,
        system_message_email: <VALUE>,
    }
}
```

File: db/migrate/20190125123107_create_user_system_notifications.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply db/migrate/20190125123107_create_user_system_notifications.rb]: We've agreed to implement it your way, just add the suffix.

**Comment by Ducz0r** [comment]
We've discussed, and decided, about certain changes about how the synchronization works. So I'm re-assigning this back to you @urbanrotnik :slightly_smiling_face: .

File: db/schema.rb [hunk 4]
```diff
@@ -660,6 +660,22 @@
...
 
+  create_table "system_notifications", force: :cascade do |t|
+    t.string "title"
+    t.text "description"
+    t.string "modal_title"
+    t.text "modal_body"
+    t.boolean "show_on_login", default: false
+    t.datetime "source_created_at"
+    t.bigint "source_id"
+    t.datetime "last_time_changed_at"
```
Comment (Ducz0r) [inline db/schema.rb]: Should this field be `null: false`?

File: spec/factories/system_notifications.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :system_notification do
+    sequence(:title) { |n| "System notification #{n}" }
+    description { Faker::ChuckNorris.fact }
+    modal_title { Faker::Name.first_name }
+    modal_body { Faker::Lorem.paragraphs(4).map { |pr| "<p>#{pr}</p>" }.join }
+    source_created_at { Faker::Time.between(3.days.ago, Date.today) }
+    source_id { Faker::Number.between(1, 1000) }
```
Comment (Ducz0r) [inline spec/factories/system_notifications.rb]: Probably we need to update the factories as well?

File: spec/models/system_notification_spec.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    describe '#source_id' do
+      it { is_expected.to validate_presence_of(:source_id) }
+    end
+
+    describe '#source_created_at' do
+      it { is_expected.to validate_presence_of(:source_created_at) }
+    end
+  end
+
```
Comment (Ducz0r) [inline spec/models/system_notification_spec.rb]: If we make it non-nullable, this also needs to be updated.

File: db/schema.rb (hunk 4 above)
Comment (urbanrotnik) [reply db/schema.rb]: I was thinking about this...
Do we see any benefits if the column is nullable? 
I thought to make it nullable for old migrated notifications, but not make sense or any advantage... 

Yeap, lets go with `null: false`...

**Comment by Ducz0r** [comment]
:+1: I hope this one is good to go now!

**Pull request merged**